syncopy/tests/test_discretedata.py

```diff
@@ -453,8 +451,100 @@ def test_ed_trialsetting(self):
         with pytest.raises(SPYValueError):
             ang_dummy.definetrial(evt_dummy, pre=pre, post=post, trigger=1)
 
+class TestWaveform():
+
+    def test_waveform_invalid_set(self):
+        """Sets invalid waveform for data: dimension mismatch"""
+        spiked = SpikeData(data=np.ones((2, 3), dtype=int), samplerate=10)
+        assert spiked.data.shape == (2, 3,)
+        with pytest.raises(SPYValueError, match="wrong size waveform"):
+            spiked.waveform = np.ones((3, 3), dtype=int)
+
+    def test_waveform_invalid_set_emptydata(self):
+        """Tries to set waveform without any data."""
+        spiked = SpikeData()
+        with pytest.raises(SPYValueError, match="Please assign data first"):
+            spiked.waveform = np.ones((3, 3), dtype=int)
+
+    def test_waveform_valid_set(self):
+        """Sets waveform in a correct way."""
+        spiked = SpikeData(data=np.ones((2, 3), dtype=int), samplerate=10)
+
+        assert not spiked._is_empty()
+        assert spiked.data.shape == (2, 3,)
+        assert type(spiked.data) == h5py.Dataset
+        assert spiked._get_backing_hdf5_file_handle() is not None
+        spiked.waveform = np.ones((2, 3), dtype=int)
+        assert "waveform" in spiked._hdfFileDatasetProperties
+        assert spiked.waveform.shape == (2, 3,)
+
+    def test_waveform_valid_set_with_None(self):
+        """Sets waveform to None, which is valid."""
+        spiked = SpikeData(data=np.ones((2, 3), dtype=int), samplerate=10)
+        assert not spiked._is_empty()
+        assert spiked.data.shape == (2, 3,)
+        spiked.waveform = np.ones((2, 3), dtype=int)
+        assert spiked.waveform.shape == (2, 3,)
+        spiked.waveform = None
+        assert spiked.waveform is None
+        # try to set again
+        spiked.waveform = np.ones((2, 3), dtype=int)
+        assert spiked.waveform.shape == (2, 3,)
+
+
+    def test_waveform_selection_trial(self):
```

Comment:
Strictly speaking, also more involved channel/unit selections could be tested here.. however as this is all nicely separated and downstream of the standard selection machinery (using the same `_preview_trial.idx` and so on) it's not needed. I also just did it in an interpreter and works as expected!

Reply:
yes, I also considered that, but I figured it would be kind of a duplicate test, because of the usage of `_preview_trial`. One thing that could be wrong is the index used as a parameter to `_preview_trial`, of course.